### PR TITLE
Use Google Blockly cross-tab copy and paste plugin

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR removes the [code-dot-org implementation ](https://github.com/code-dot-org/code-dot-org/pull/45862) of cross-tab copy and paste using context menu and keyboard shortcuts in Google Blockly labs like Poetry and re-adds this functionality with this [plugin](https://github.com/google/blockly-samples/tree/master/plugins/cross-tab-copy-paste). I removed our implementation by removing the `shortcut.js` file (yay!) and deleting `registerBlockCopyToStorage` and `registerBlockPasteFromStorage` in `contextMenu.js`. Note that the plug-in already implements the keyboard shortcuts [here](https://github.com/google/blockly-samples/blob/f22cf2e5416b2926dcc684149bc269ec05b57cea/plugins/cross-tab-copy-paste/src/index.js#L137-L274).

We couldn't use this plugin until we migrated to at least Google Blockly v7. We moved to Google Blockly v8 (skipping over v7) on June 28 with [this PR](https://github.com/code-dot-org/code-dot-org/pull/46995). We are now [at v9](https://github.com/code-dot-org/code-dot-org/pull/50172). However, there were inconsistencies in the preconditions for the 'copy' action as described in the 'Follow-up work' section below. Thus, we waited for the plug-in bug fix before merging this PR. We are now at the `@blockly/plugin-cross-tab-copy-paste` version ^2.0.7.

 - screencast videos after the PR updates:

Screencast of copy and paste using context menu and keyboard shortcuts:

https://user-images.githubusercontent.com/107423305/200923169-7374b4e7-5978-49d9-acb3-ad994d789f1e.mp4

Screencast of cut and paste using the keyboard shortcut (cut is not an option with the context menu):

https://user-images.githubusercontent.com/107423305/200923179-30fe2a8a-e065-4a6c-8cf8-a83d1fcdbde5.mp4

A note that when I first implemented this plug-in, the 'cut' action was not working properly. It removed a block and its next blocks but only pasted the block. 

Screencast video of example:
https://user-images.githubusercontent.com/107423305/199839701-aba3d248-9d63-4612-896c-7704de9f920f.mp4

The first thought was that there was a bug in the plugin, but the bug was actually in code-dot-org's overriding of the `dispose` function. This bug was fixed by [this pr](https://github.com/code-dot-org/code-dot-org/pull/49073). 

## Links
[jira ticket - [Google Blockly] Add plug-in for cross-tab Copy/Paste](https://codedotorg.atlassian.net/browse/SL-21)

## Testing story
I tested locally (including on levelbuilder in start mode). 

## Follow-up work
@mikeharv noticed that there were inconsistencies in the preconditions for the 'copy' action. The context menu option for 'copy' is set to always be enabled while the shortcut for copy has preconditions including that a block is deletable and movable. Core Blockly does not allow immovable nor undeletable blocks to be copied. Thus, Mike H submitted [this PR](https://github.com/google/blockly-samples/pull/1371) to fix this inconsistency. 

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/addons/contextMenuAndShortcut.js
Comment (mikeharv) [inline apps/src/blockly/addons/contextMenuAndShortcut.js]: This precondition is important because we want to prevent users from copying blocks they immovable/undeletable.

This is the behavior on levelbuilder today that we want to preserve:

https://user-images.githubusercontent.com/43474485/200342753-508738a7-25b5-4952-85b7-20740dd1aed4.mp4

To accomplish this, I think we'll need to override the `preconditionFn` function here: https://github.com/google/blockly-samples/blob/304a875648bac4c7d78747a298dc8b2a1d28ba5b/plugins/cross-tab-copy-paste/src/index.js#L67-L70

File: apps/src/blockly/addons/shortcut.js
Comment (mikeharv) [inline apps/src/blockly/addons/shortcut.js:17]: Note that this removal is fine, because the plugin already does this by default: https://github.com/google/blockly-samples/blob/304a875648bac4c7d78747a298dc8b2a1d28ba5b/plugins/cross-tab-copy-paste/src/index.js#L184-L185

**Review by mikeharv — COMMENTED** [review]
Great work, Alice! We need to resolve two issues here before we can merge this. First is the bug you brought up, which I believe resides in the plugin. Second is that the copy command needs to be disabled if the block shouldn't be copied. We can check with Maribeth whether she thinks this should be part of the plugin's core functionality or something we should override.

File: apps/src/blockly/addons/contextMenuAndShortcut.js
Comment (fisher-alice) [reply apps/src/blockly/addons/contextMenuAndShortcut.js]: I compared the  copying behavior with what's currently on Levelbuilder and how it's implemented in the plugin:
As you described, currently users are unable to copy and paste unmovable or undeletable blocks. Here's an example with an undeletable block:

https://user-images.githubusercontent.com/107423305/200419197-32275173-2e1e-47ae-b46e-3b5ef1df6b2a.mp4

With the plugin, users are able to copy and paste an unmovable or undeletable block, but when it's pasted, the settings for the new block are that it's movable and deletable. Here's an example with an undeletable block on this branch:

https://user-images.githubusercontent.com/107423305/200419461-c13a621d-9035-4ea8-9dce-29ebdd67c07f.mp4

I wonder if the behavior using the plugin implementation is actually more desirable for a user.

Comment (fisher-alice) [reply apps/src/blockly/addons/contextMenuAndShortcut.js]: After consulting with @maribethb, we agreed to follow the core Blockly behavior and not allow copying of immovable and undeletable blocks. @mikeharv 's [PR](https://github.com/google/blockly-samples/pull/1371) will fix the current inconsistency with the 'Copy' action.

File: apps/src/blockly/addons/contextMenuAndShortcut.js
Comment (mikeharv) [inline apps/src/blockly/addons/contextMenuAndShortcut.js]: Why do we need these?

File: apps/src/blockly/addons/contextMenuAndShortcut.js
Comment (mikeharv) [inline apps/src/blockly/addons/contextMenuAndShortcut.js]: Is this line the only thing here that has to do with shortcuts? Rather than renaming this file, I wonder if it would be cleaner to import and initialize the plugin elsewhere (maybe right in the wrapper), and leave this file exclusively for _our_ custom context menu options.

File: apps/src/blockly/googleBlocklyWrapper.js
Comment (mikeharv) [inline apps/src/blockly/googleBlocklyWrapper.js]: Could we initialize the plugin here and leave this line as `registerAllContextMenuItems()`?

File: apps/src/blockly/addons/contextMenuAndShortcut.js
Comment (mikeharv) [inline apps/src/blockly/addons/contextMenuAndShortcut.js]: `registerAllContextMenuItems` still only includes context menu options (and no shortcuts). `registerAllContextMenuShortcutItems` feels over-complex and a little inaccurate give there aren't any shortcuts being exported.

**Review by mikeharv — COMMENTED** [review]
I wonder if we should plan to merge this now or wait for the plugin to get updated with our change: 
- https://github.com/google/blockly-samples/pull/1371

Any chance you've already thought through all the possible ways we could ship this change along with the change to `dispose`?

How will copy/cut/paste change if we:
- ship both changes now
- just ship the change to `dispose` now but wait for the plugin to be updated before merging this
- wait to ship either change until the plugin is updated

File: apps/src/blockly/addons/contextMenuAndShortcut.js
Comment (fisher-alice) [reply apps/src/blockly/addons/contextMenuAndShortcut.js]: Ahhh...thought we need the assignment or we would get 'CROSS_TAB_COPY' but now see 'Copy' is the default. Thanks! Will remove.

File: apps/src/blockly/addons/contextMenuAndShortcut.js
Comment (fisher-alice) [reply apps/src/blockly/addons/contextMenuAndShortcut.js]: Sure - that sounds good to me!

File: apps/src/blockly/googleBlocklyWrapper.js
Comment (fisher-alice) [reply apps/src/blockly/googleBlocklyWrapper.js]: Yes - I agree this is a better place to initialize the plugin - thanks!

File: apps/src/blockly/addons/contextMenuAndShortcut.js
Comment (fisher-alice) [reply apps/src/blockly/addons/contextMenuAndShortcut.js]: That definitely makes sense -just pushed the changes - thanks!

File: apps/src/blockly/googleBlocklyWrapper.js
Comment (mikeharv) [inline apps/src/blockly/googleBlocklyWrapper.js:50]: Wow! The list of changed files is so much easier to read now with these last couple of commits. Great work, @fisher-alice!

**Review by mikeharv — COMMENTED** [review]
This is looking good! Let's sync offline to determine the best way to get these changes out. I want us to be mindful about the update we submitted to blockly-samples as well our plans for the v9 bump. Thanks so much for getting all this staged. It's especially great that this all happened because it revealed the issue with the `dispose` override. That's been hanging out on out Google Blockly labs all year and I'm so glad we caught before we flipped Dance Party. 🎉

**Comment by fisher-alice** [comment]
> I wonder if we should plan to merge this now or wait for the plugin to get updated with our change:
> 
> * [google/blockly-samples#1371](https://github.com/google/blockly-samples/pull/1371)
> 
> Any chance you've already thought through all the possible ways we could ship this change along with the change to `dispose`?
> 
> How will copy/cut/paste change if we:
> 
> * ship both changes now
> * just ship the change to `dispose` now but wait for the plugin to be updated before merging this
> * wait to ship either change until the plugin is updated

Since the `dispose` fix is already merged, we now have the following options:
1. Ship this change now.
2. Wait to ship until plugin is updated.

If we ship the change now, the issue is that undeletable or unmovable blocks will be able to be copied using the context menu. When these blocks are pasted, they are NOT undeletable nor immovable because these properties are not included in the serialization of the block. But as discussed above, we want to match the core Blockly behavior of users not being able to copy undeletable or unmovable blocks. 

If we don't ship the plugin now, then there is a small bug in the current implementation of and cut/paste- 'cut' removes a block but 'paste' includes the next block. The 'copy' action also includes the next blocks. Although there is a manual workaround (user can just deleted the extra next blocks) the issue should be addressed. 

UPDATE: [This PR](https://github.com/code-dot-org/code-dot-org/pull/49087)  was merged on 11/14/22 and fixes the issue of next blocks being included in the 'cut' and 'copy' actions. The 'next' blocks are now ignored!

File: apps/src/blockly/googleBlocklyWrapper.js
Comment (fisher-alice) [reply apps/src/blockly/googleBlocklyWrapper.js:50]: It was definitely a team effort - thanks!

**Review by mikeharv — APPROVED** [review]
Look at all that beautiful red! 🧹 

Thanks for carrying this over the finish line.

**Pull request merged**
